### ISSUE TITLE
Extremely high memory consumption

### ISSUE THREAD

**Issue by jdicami** [Nozbe/WatermelonDB#948]
I currently have an issue where the initial sync of my application is using massive amount memory (1-2gb) without feeing when the sync is complete. This is causing memory pressure warnings and sometimes crashing. Essentially we are trying sync around 70k+ rows which explains why the memory usage is so high, by why is the memory not being free'd. 

Our endpoint is paginated so I tweaked the default synchronized slightly so that the batching is done per page (5k records per page). 

```
        for await (let page of changes) {
            const adapter = new SyncAdapter(database)

            let {
                changes: remoteChanges, 
                timestamp 
            } = await adapter.getChangeRecords(page)

            newLastPulledAt = timestamp

            await action.subAction(() =>
                applyRemoteChanges(
                    database,
                    remoteChanges,
                    sendCreatedAsUpdated,
                    log,
                    conflictResolver,
                    _unsafeBatchPerCollection,
                ),
            )

            adapter.clear()

            remoteChanges = null
        }    

```

This helps a little, but the issue remains where the memory is not being freed when sync is complete. Is watermelon caching or retaining the data?

**Comment by radex** [comment]
how large are your rows? 1-2GB for 70k is a lot.

yes, unfortunately the mass-insert logic in sync is not highly refined (yet!), so all the records stick around in the in-JS cache.

for quick solutions, see if you can easily rewire your app's logic so that after this mass-insert, you get rid of your `Database` instance, and instantiate a new one. If you don't hold onto any references to the former, it should get garbage collected.

alternatively, you could dig around Watermelon's code to clear the caches, but you'd also have to mess around with native code… probably easier to create a new Database

**Comment by jdicami** [comment]
Thank you for the suggestion of recreating the instance, which was helpful. The rows are unusually large due to field containing large amounts of JSON and unfortunately the source of the data is out of my control. What I ended up doing is having the app download a prepopulated database from a snapshot in time and then hacking a method (in native) in watermelon to Attach the snapshot sqlite database and copy the tables over.

**Comment by radex** [comment]
Yes, that sounds like a very good solution for this problem.

**Comment by lincolnbf** [comment]
@radex Thank you, youre the best, you saved my day!!!!

**Comment by lincolnbf** [comment]
I am having a problem with this approach. I'm trying to reset the database instance this way, after the mass insert. But when i try to access the db after the reset, i have the RecordId error. 

`export let database = new Database({
  adapter,
  modelClasses: models,
});

export function resetDbInstance() {
  database = new Database({
    adapter,
    modelClasses: models,
  });
}`

**Comment by nullgr4vity** [comment]
hi guys, 

I also experiencing similar issue with memory consumption after sync. My situation is a little bit different as my first synchronisation is around 120MB JSON (and growing). It was too much for one time sync, so I divided it into ~30 smaller pages, where biggest page picked to around 15mb. Then I run synchronisation in loop, each iteration get another page from server. Something like in code below:

```
let hasNext = true;
while (hasNext) {

      await synchronize({
        database,
        log,
        pullChanges: async (args) => {
          const {
            changes,
            nextPageAt = null,
          } = await pullChangesRoutine({
            ...args,
          });

          hasNext = nextPageAt !== null;

          return { changes, timestamp: nextPageAt };
        },
        pushChanges: (args) =>
          pushChangesRoutine({ ...args }),
      });
}
```

However each synchronisation doesn't release memory, at the end of the process my app consumed around 600 mb memory consumed. 

I tried the idea which was discussed above - release connection to db and create new one. It was better but still memory consumptions was around 450mb. Just to add - synchronization is a "first thing ever" in app, no other references.

When testing in instruments, I see that batchJSON() consumed a lot of memory. Is this method could be an issue ?
